core/modules/layout/plugins/context/layout_context.inc

```diff
@@ -5,6 +5,39 @@
  * Class that holds information relating to a layouts context.
  */
 abstract class LayoutContext extends LayoutHandler {
+
+  /**
+   * Declares a LayoutContext to be a manually specified custom context.
+   */
+  const USAGE_TYPE_CUSTOM = 1;
+
+  /**
+   * Declares a LayoutContext to be provided based on a menu item or path.
+   */
+  const USAGE_TYPE_MENU = 2;
+
+  /**
+   * Declares a LayoutContext to be a system-wide context
+   *
+   * Examples include the "current_user" context and "overrides_path" context.
+   */
+  const USAGE_TYPE_SYSTEM = 4;
+
+  /**
+   * Declares a LayoutContext is provided by a relationship to another context.
+   */
+  const USAGE_TYPE_RELATIONSHIP = 8;
```

Comment:
For the sake of curiosity, can someone explain to me why we are using what seem to me exponentially increasing numbers for these constants? What is the trick we're using them for?

Reply:
They are for bitwise math operations, where binary numbers are used to indicate one "flag". These numbers in binary correlate to the following in binary:

```
1 = 0001
2 = 0010
4 = 0100
8 = 1000
```

Each number results in a binary number with a single 1 digit and the rest zeros.

Then you can use multiple flags to request multiple types of layout usages by using [bitwise math operators](https://www.php.net/manual/en/language.operators.bitwise.php), like this:

```
USAGE_TYPE_CUSTOM | USAGE_TYPE_MENU
```

Which turns into `0011` (both 1 and 2). That means you can ask for multiple flags in a single argument instead of passing in several arguments, or having a bunch of IF statements that work on strings. This same pattern is used in the menu system in Backdrop, and is common in PHP functions, for example [`preg_replace_all()`](https://www.php.net/manual/en/function.preg-match-all.php).

Reply:
Thanks @quicksketch 🙏🏼